core/capabilities/remote/target/receiver.go

```diff
@@ -0,0 +1,129 @@
+	"github.com/smartcontractkit/chainlink/v2/core/capabilities/remote/target/request"
+	"github.com/smartcontractkit/chainlink/v2/core/capabilities/remote/types"
+	p2ptypes "github.com/smartcontractkit/chainlink/v2/core/services/p2p/types"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+)
+
+// receiver manages all external users of a local target capability.
+// Its responsibilities are:
+//  1. Manage requests from external nodes executing the target capability once sufficient requests are received.
+//  2. Send out responses produced by an underlying capability to all requesters.
+//
+// receiver communicates with corresponding caller on remote nodes.
+type receiver struct {
+	lggr         logger.Logger
+	peerID       p2ptypes.PeerID
+	underlying   commoncap.TargetCapability
+	capInfo      commoncap.CapabilityInfo
+	localDonInfo capabilities.DON
+	workflowDONs map[string]commoncap.DON
+	dispatcher   types.Dispatcher
+
+	requestIDToRequest map[string]*request.ReceiverRequest
+	requestTimeout     time.Duration
+
+	receiveLock sync.Mutex
+}
+
+var _ types.Receiver = &receiver{}
+
+func NewReceiver(ctx context.Context, lggr logger.Logger, peerID p2ptypes.PeerID, underlying commoncap.TargetCapability, capInfo commoncap.CapabilityInfo, localDonInfo capabilities.DON,
+	workflowDONs map[string]commoncap.DON, dispatcher types.Dispatcher, requestTimeout time.Duration) *receiver {
+	r := &receiver{
+		underlying:   underlying,
+		peerID:       peerID,
+		capInfo:      capInfo,
+		localDonInfo: localDonInfo,
+		workflowDONs: workflowDONs,
+		dispatcher:   dispatcher,
+
+		requestIDToRequest: map[string]*request.ReceiverRequest{},
+		requestTimeout:     requestTimeout,
+
+		lggr: lggr,
+	}
+
+	go func() {
+		ticker := time.NewTicker(requestTimeout)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				r.ExpireRequests()
+			}
+		}
+	}()
+
+	return r
+}
+
+func (r *receiver) ExpireRequests() {
+	r.receiveLock.Lock()
+	defer r.receiveLock.Unlock()
+
+	for requestID, executeReq := range r.requestIDToRequest {
+		if executeReq.Expired() {
+			err := executeReq.Cancel(types.Error_TIMEOUT, "request expired")
+			if err != nil {
+				r.lggr.Errorw("failed to cancel request", "request", executeReq, "err", err)
+			}
+			delete(r.requestIDToRequest, requestID)
+		}
+	}
+}
+
+func (r *receiver) Receive(msg *types.MessageBody) {
+	r.receiveLock.Lock()
+	defer r.receiveLock.Unlock()
+	// TODO should the dispatcher be passing in a context?
+	ctx := context.Background()
+
+	if msg.Method != types.MethodExecute {
+		r.lggr.Errorw("received request for unsupported method type", "method", msg.Method)
+		return
+	}
+
+	// A request is uniquely identified by the message id and the hash of the payload to prevent a malicious
+	// actor from sending a different payload with the same message id
```

Comment:
I'm still not a fan of this. I'm worried that it'll make debugging harder. IMO a more natural logic will be to cluster messages based on WorkflowExecutionID and then pick F+1 identical payloads once we have them. Your logic achieves the same outcome so I'm not going to block it.